src/classes/emitter.ts

```diff
@@ -1,41 +1,72 @@
 // deno-lint-ignore-file no-explicit-any
-export type Listener = (this: Emitter, ...args: any[]) => void;
+export type Listener<
+	TThis = any,
+	TArgs extends any[] = any[],
+> = (this: TThis, ...args: TArgs) => void;
 
-export default class Emitter {
-	#events: Record<string, Listener[]> = {};
+/**
+ * A record mapping event's names to its listener arguments.
+ */
+export type EmitterEvents<
+	TEventName extends string = string,
+	TEventArguments extends any[] = any[],
+> = {
+	[eventName in TEventName]: TEventArguments;
+};
 
-	on(e: string, func: Listener) {
-		if (typeof this.#events[e] !== "object") {
+export default class Emitter<
+	TEmitterEvent extends EmitterEvents = EmitterEvents,
+> {
+	#events: {
+		[TKey in keyof TEmitterEvent]?: Listener<this, TEmitterEvent[TKey]>[];
+	} = {};
+
+	on<TKey extends keyof TEmitterEvent>(
+		e: TKey,
+		func: Listener<this, TEmitterEvent[TKey]>,
+	) {
+		if (!Array.isArray(this.#events[e])) {
```

Comment:
Why this check either it is `undefined` or an Array. (empty arrays doesn't matter)

Reply:
Yup, that's what I thought, too. But somehow with this Deno/Node setup it would still complain... hence I added the `!`